### PR TITLE
[Autoscaler] Match autoscaler image to Ray head image for Ray >= 2.0.0

### PR DESCRIPTION
**Pull request by DmitriGekhtman**
## Why are these changes needed?

This PR updates the logic that selects a default autoscaler image.

For Ray versions at least 2.0.0, use the same image for the autoscaler as for the Ray container.
This eliminates the possibility of autoscaler/Ray incompatibility and reduces docker pull time.

For earlier Ray versions, use `rayproject/ray:2.0.0` to guarantee up-to-date autoscaler functionality.
As of the Ray 2.0.0 branch cut earlier today, an image tagged `rayproject/ray:2.0.0` exists on Dockerhub.
Until the official Ray release in two weeks, this image is unofficial and its actual contents are a moving target -- but I think we can live with the inconsistency in the short term. (I'm open to other opinions on this choice.) 

## Related issue number

Closes https://github.com/ray-project/kuberay/issues/360

## Checks

- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [ ] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by Jeffwan** [comment]
so `rayproject/ray:2.0.0` will be overridden once  the office image is out?

**Comment by DmitriGekhtman** [comment]
> so `rayproject/ray:2.0.0` will be overridden once the office image is out?

`rayproject/ray:2.0.0` is updated each time we push to the Ray 2.0.0 release branch

The "official image is out" after the last commit is made to the release branch and we announce the release.

Maybe not the best way of doing things but that's the way it is at the moment.

File: ray-operator/controllers/ray/common/constant.go [hunk 1]
```diff
@@ -92,6 +92,9 @@ const (
 	// Ray health check related configurations
 	RayAgentRayletHealthPath  = "api/local_raylet_healthz"
 	RayDashboardGCSHealthPath = "api/gcs_healthz"
+
+	// Default autoscaler image when running Ray at versions older than 2.0.0
+	FallbackDefaultAutoscalerImage = "rayproject/ray:2.0.0"
```
Comment (DmitriGekhtman) [inline ray-operator/controllers/ray/common/constant.go:97]: This image exists but is unofficial at the moment.
Its contents may change slightly over the course of the Ray 2.0.0 release process.

File: ray-operator/controllers/ray/common/pod.go [hunk 2]
```diff
@@ -139,6 +143,44 @@ func DefaultHeadPodTemplate(instance rayiov1alpha1.RayCluster, headSpec rayiov1a
...
+	}
+}
+
+// Determine if autoscaler support is stable in the given rayVersion.
+// Return false exactly when the major version is successfully parsed and less than 2.
+// Example rayVersion inputs that return true: "2.0.0", "2.0", "2", "2.0.0rc1", "nightly", "latest", "unknown".
+// Example inputs that return false: "1.13.0", "1.12", "1".
+func autoscalerSupportIsStable(rayVersion string) bool {
+	// Try to determine major version by extracting everything that comes before the first "."
+	firstDotIndex := strings.Index(rayVersion, ".")
```
Comment (Jeffwan) [inline ray-operator/controllers/ray/common/pod.go:163]: minor: In our downstream code, we have similar version specific configuration. for those cases, we use `hashicorp/go-version` to directly compare versions if the version follow semantic versioning. 

```go
import "github.com/hashicorp/go-version"

// LessThanVersion tests if version1 is less than version2.
func LessThanVersion(version1 string, version2 string) (bool, error) {
	v1, err := version.NewVersion(version1)
	if err != nil {
		return false, err
	}
	v2, err := version.NewVersion(version2)
	if err != nil {
		return false, err
	}

	return v1.LessThan(v2), nil
}
```

**Review by Jeffwan — APPROVED** [review]
Overall looks good to me and I leave a minor comment which may help simplify the version comparison

**Comment by Jeffwan** [comment]
#424 is merged to help improve test stability. You can rebase the change to see if nightly version pass

File: ray-operator/controllers/ray/common/pod.go (hunk 2 above)
Comment (DmitriGekhtman) [reply ray-operator/controllers/ray/common/pod.go:163]: Thanks for the tip! I'll keep it in mind to use that library for version comparisons in the future.

For now, I prefer to implement the logic directly, as I'd like to be super explicit with what we're doing with ill-formatted version strings.

**Comment by Jeffwan** [comment]
Let's merge this one and it's time to cut rc.0 release. If other reviewers have further feedback, feel free to leave it here.

**Pull request merged**